src/ott/solvers/linear/univariate.py

```diff
@@ -318,29 +331,120 @@ def quantile_distance(
   y_cdf_inv = all_values_sorted[i_y_cdf_inv]
 
   diff_q = jnp.diff(quantile_levels)
-  cost = jnp.sum(
-      jax.vmap(cost_fn.h)(y_cdf_inv[1:, None] - x_cdf_inv[1:, None]) * diff_q
+  successive_costs = jax.vmap(cost_fn.h)(
+      x_cdf_inv[1:, None] - y_cdf_inv[1:, None]
   )
-  if not return_transport:
-    return cost, None, None
+  cost = jnp.sum(successive_costs * diff_q)
 
-  n = x.shape[0]
+  if return_transport:
+    n = x.shape[0]
+
+    i_in_sorted_x_of_quantile = all_values_sorter[i_x_cdf_inv] % n
+    i_in_sorted_y_of_quantile = all_values_sorter[i_y_cdf_inv] - n
+
+    orig_i = i_x[i_in_sorted_x_of_quantile][1:]
+    orig_j = i_y[i_in_sorted_y_of_quantile][1:]
+    paired_indices, mass_paired_indices = jnp.stack([orig_i, orig_j]), diff_q
+  else:
+    paired_indices, mass_paired_indices = None, None
+
+  return cost, paired_indices, mass_paired_indices
+
+
+# TODO(michalk8): better naming
+def north_west_distance(
```

Comment:
I would suggest `north_west_corner_rule`  (e.g. https://arxiv.org/pdf/1803.00567 3.4.2). There it's in the primal, here we're running it in the dual.